samples/go/photo-dedup/job/hashjob.go

```diff
@@ -0,0 +1,143 @@
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package job
+
+import (
+	"fmt"
+	"image"
+	_ "image/gif"
+	_ "image/jpeg"
+	_ "image/png"
+	"math"
+	"net/http"
+	"os"
+	"runtime"
+	"sync"
+
+	"github.com/attic-labs/noms/go/datas"
+	"github.com/attic-labs/noms/go/types"
+	"github.com/attic-labs/noms/go/util/status"
+	"github.com/attic-labs/noms/go/walk"
+	"github.com/attic-labs/noms/samples/go/photo-dedup/dhash"
+	"github.com/attic-labs/noms/samples/go/photo-dedup/model"
+)
+
+// HashPhotosJob adds a dhash field to every photo in photoSets and commits them to
+// a new Photo set in outDS. The dhash is used in turn by the DeduplicatePhotosJob
+// to group similar photos into PhotoGroups.
+func HashPhotosJob(db datas.Database, photoSets []types.Value, outDS datas.Dataset) error {
+	return commitHashedPhotos(db, outDS, hashPhotos(db, photoSets))
+}
+
+// hashPhotos adds a dhash to each photo in photoSets and delivers them on the returned channel
+func hashPhotos(db datas.Database, photoSets []types.Value) <-chan types.Struct {
+	photosToHashCh := make(chan model.Photo)
+	hashedPhotosCh := make(chan types.Struct)
+
+	go func() {
+		for _, set := range photoSets {
+			walk.WalkValues(set, db, func(cv types.Value) (stop bool) {
+				if photo, ok := model.UnmarshalPhoto(cv); ok {
+					photosToHashCh <- photo
+				}
+				return false
+			})
+		}
+		close(photosToHashCh)
+	}()
+
+	fmt.Print("Downloading and hashing photos...\n")
+	numWorkers := runtime.NumCPU() * 4
+	var wg sync.WaitGroup
+	wg.Add(numWorkers)
+	for i := 0; i < numWorkers; i++ {
+		go func() {
+			defer wg.Done()
+			for photo := range photosToHashCh {
+				hashed, err := addHashToPhoto(photo)
+				if err != nil {
+					fmt.Fprintf(os.Stderr, "\nSkipping: %s\n", err)
+					hashedPhotosCh <- photo.Marshal()
```

Comment:
I think so. In this case, there are failures that occur because of formats we don't handle, images that aren't reachable and possibly rate limiting. It makes sense to skip past them but we don't want to lose track of the fact that they happened. I think we need a mechanism for logging errors and warning in plugins in general.